fish_completions.go

```diff
@@ -43,16 +43,13 @@ function __%[1]s_perform_completion
     end
     __%[1]s_debug "emptyArg: $emptyArg"
 
-    if not type -q "$args[1]"
-        # This can happen when "complete --do-complete %[2]s" is called when running this script.
-        __%[1]s_debug "Cannot find $args[1]. No completions."
-        return
-    end
-
     set requestComp "$args[1] %[3]s $args[2..-1] $emptyArg"
     __%[1]s_debug "Calling $requestComp"
 
-    set results (eval $requestComp 2> /dev/null)
+    # Call the command as a sub-shell so that we can redirect any errors
+    # For example, if $requestComp has an unmatched quote
+    # https://github.com/spf13/cobra/issues/1214
+    set results (fish -c "$requestComp" 2> /dev/null)
```

Comment:
Hmm this is interesting. If you didn't include the token at the cursor this wouldn't be a problem.

Usually completions want to use `commandline -opc`, which gives all words left of the cursor, which always tokenize successfully.

It also restricts the tokens to  the current process (`-p`). Observe bind running
```
bind \eg 'commandline -c'
```
and then type `echo foo; echo bar` and <kbd>Ctrl>/kbd>+<kbd>g</kbd>.
So `-p` is usually correct for completions.

The `-o` flag is useful because it already does word splitting, then there is no need for `string split " "`.

Reply:
OK I've played around with this some more, and found a version that should tick all boxes, while not needing a separate fish process. Here's what I would do:

```fish
function __%[1]s_perform_completion
    set -l args (commandline -opc)
    set -l lastArg (commandline -ct)
    set -l requestComp $args[1] %[3]s $args[2..-1] (string escape -- $lastArg)
    if test -d $args[1]
        return 1
    end
    set -l results (eval $requestComp 2> /dev/null)
```

- `set -l` is a minor detail: if a user defines a universal variable with `set -U x foo`, then `set x` inside this function would overwrite the user's variable. This is an ugly edge case in fish which should probably be changed, but for now we can force variables to be local by use `set -l` when they are used for the first time in a function.
- `(commandline -ct)` gives us the raw token at the cursor, including any unclosed quotes. To avoid https://github.com/spf13/cobra/issues/1214, we can simply escape it. This means that variables in the token at the cursor will not be evaluated, but that's a general issue with user-defined fish completions, and I don't think it's an important problem.
- If the command is a directory I got  some errors (trying to complete `./testprog `), so I added a check for that.

Reply:
Thanks!  I'll take the time to look into this as soon as I can and get back to you.

Reply:
@krobelus I finally had a chance to try your suggestion above.  I really love the fact that we no longer have to explicitly check for an empty lastArg thanks to the escaping.  Thanks for all your help!

I will put `-l` everywhere there is a `set` of a new variable.

I've also slightly modified your suggestion to do the escaping right away when calculating `lastArg`; if find it easier to read.